components/gatekeeper/auth/AuthServer.go

```diff
@@ -0,0 +1,187 @@
+// limitations under the License.
+
+package auth
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/kubeflow/kubeflow/components/gatekeeper/cmd/gatekeeper/options"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/bcrypt"
+	"math/rand"
+	"net/http"
+	"path"
+	"strings"
+	"sync"
+	"time"
+)
+
+// start with easy case: each server struct only has one valid pair of u/p
+type authServer struct {
+	username	string
+	// password bcrypt hash
+	pwhash      string
+	// authorized cookies and their expire time (12 hour by default)
+	cookies 	map[string]time.Time
+	serverMux   sync.Mutex
+}
+
+const CookieName = "KUBEFLOW-AUTH-KEY"
+const LoginPagePath = "kflogin"
+const LoginPageHeader = "x-from-login"
+
+func NewAuthServer(opt *options.ServerOption) *authServer {
+	data, err := base64.StdEncoding.DecodeString(opt.Pwhash)
+	if err != nil {
+		log.Fatal("error:", err)
+	}
+	server := &authServer{
+		username: opt.Username,
+		pwhash: string(data),
+		cookies: make(map[string]time.Time),
+	}
+	return server
+}
+
+// Default auth check service
+func (s *authServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Infof("Path check, url: %v, path: %v", r.URL, r.URL.Path)
+	// login page open to everyone; all other path requires auth with Password or cookie
+	if strings.HasPrefix(r.URL.Path, "/" + LoginPagePath) || s.authCookie(r) == true {
+		// Handle request from login page
+		if r.Header.Get(LoginPageHeader) != "" {
+			w.WriteHeader(http.StatusResetContent)
+			w.Write([]byte(http.StatusText(http.StatusResetContent)))
+			return
+		}
+		// Allow browser request
+		log.Infof("Allow browser request")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(http.StatusText(http.StatusOK)))
+		return
+	}
+
+	if s.authpwd(r) == true {
+		log.Infof("P/W passed")
+		// Handle request from login page
+		if r.Header.Get(LoginPageHeader) != "" {
+			s.setCookieAndReset(w, r)
+			return
+		}
+		// Allow requst from API call
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(http.StatusText(http.StatusOK)))
+		return
+	}
+	log.Infof("Unauthorized, redirect to %v", "https://" + path.Join(r.Host, LoginPagePath))
+
+	// redirect to login page
+	s.redirectToLogin(w, r)
+}
+
+// auth with basic pw
+func (s *authServer) authpwd(r *http.Request) bool {
+	auth := r.Header.Get("Authorization")
+	if !strings.HasPrefix(strings.ToLower(auth), "basic ") {
+		return false
+	}
+
+	upBytes, err := base64.StdEncoding.DecodeString(auth[len("basic "):])
+	if err != nil {
+		return false
+	}
+
+	namepw := strings.Split(string(upBytes), ":")
+
+	if len(namepw) != 2 {
+		return false
+	}
+	err = bcrypt.CompareHashAndPassword([]byte(s.pwhash), []byte(namepw[1]))
+	if namepw[0] == s.username && err == nil {
+
+		return true
+	}
+	return false
+}
+
+// auth with cookie
+func (s *authServer) authCookie(r *http.Request) bool {
+	if cookie, err := r.Cookie(CookieName); err == nil {
+		if val, ok := s.cookies[cookie.Value]; ok {
```

Comment:
What prevents someone from spoofing a cookie?

Reply:
```SameSite: http.SameSiteStrictMode```
We restrict the cookie so can only be used for same site.
Should we encrypt value as well?

Reply:
You mean with bcrypt? So we only keep track of the encrypted value?

Reply:
Yes, server side only keep encrypted password hash.

Reply:
If its not too difficult I think that would be valuable; for the same reason its valuable with the password.